### ISSUE TITLE
subcommand_missing only invoked if command has defined at least one subcommand

### ISSUE THREAD

**Issue by mikeburr** [mdub/clamp#69]
This code runs as expected:

```
#!/usr/bin/env ruby

require 'clamp'

Clamp do
  subcommand 'foo', 'bar'

  def subcommand_missing(name)
    abort "would have looked for subcommand #{name} here"
  end
end
```

If you comment out the subcommand, however, you just get:

```
ERROR: too many arguments

See: 'clamptest.rb --help'
```

This appears to be due to the has_subcommands? check in parse_subcommand, but I don't understand the code well enough yet to suggest a solution.

**Comment by mdub** [comment]
Yup - the subcommand-handling functionality only kicks in if you have declared subcommands.  I would want/expect `subcommand_missing` to be invoked for a simple command where subcommands were not involved; see

* https://github.com/mdub/clamp/blob/master/spec/clamp/command_spec.rb#L769-L779

What's your use-case?

**Comment by mikeburr** [comment]
I'm building a tool where everything is a dynamically-loaded subcommand. The tool wraps the API for a rather complex backend system and provides:

- logic to automatically authenticate & do various housekeeping gorp needed to use the API
- a subclass of Clamp::Command that adds a bunch of utility methods to simplify using the API
- a dozen subcommands that do common things with the API
- the ability for users to write/share their own subcommands; each subcommand is just a Ruby file containing the guts of a dynamically-created extension of my Clamp::Command subclass

I had hacked up a small extension to Clamp v1.0 that did essentially what subcommand_missing does. It wasn't quite as elegant, but it did override has_subcommands? to always return true when the command indicated that it wanted to dynamically generate subcommands.

**Comment by mdub** [comment]
Ok. I understand what you're after, but I don't think `subcommand_missing` is the best way to get there.

@kke and I discussed auto-loading of sub commands in #61; would the solution described there work for you?

**Comment by mikeburr** [comment]
It certainly handles the load-on-use part, but I'm also aiming for discover-on-use. That is, I would like an end user to be able to write a new subcommand by simply dropping a ruby script in the right directory.

It looked like the solutions you & kke discussed were all predicated on `subcommand` calls for every possible subcommand? I toyed with the idea of iterating through my subcommand directories and making `subcommand` calls for each script, but that (a) is potentially expensive in cases where there are hundreds of subcommands, (b) complicates and/or precludes retrieving subcommands via non-file-based mechanisms that don't support enumerating content (e.g. HTTP). I realize that not having the `subcommand` calls means I also get incomplete command help, but in the hundreds of subcommands case that might not be a bad thing.

What I'm trying to do might be enough outside Clamp's mainstream use case that I just have to live with a little bit of friction.

**Comment by mdub** [comment]
Heh, you're certainly pushing the boundaries, @mikeburr.

Overriding `has_subcommands?` to return true _might_ do the trick. Or have you tried that already?

**Comment by mikeburr** [comment]
It worked until 1.1.  :-)

Here's a [gist](https://gist.github.com/mikeburr/51aa9557523c52af830f8e2c9a95313e) with my 1.0-based hacks.

**Comment by kke** [comment]
Discover on use?

Isn't that just a case of naming convention?

For example you could have a file naming convention that the filename must be the subcommand name. Or it could be something like `user-create-foo_command`  which would add it as `cmd user create foo`

A simple one for main command only would be something like:

```ruby
  def subcommand_missing(name)
    path = File.join(plugin_dir, "#{name}.rb")
    if File.exist?(path)
      require path
      Plugin.const_get(name.camelize)   # you need something to camelize strings. 
                                                             # i think it's possible to use TracePoint
                                                             # to track which class was added(?)
    else
      super(name)
    end
  end
```

If you need the subcommands to appear in helps, then i don't think you have any other choice but to require all the subcommand-plugins for current command.

And there you need something like
```ruby
class PluginCmd < Clamp::Command
  def self.inherited(where)
    where.parent_class.subcommand where.cmd, where.desc, where
  end

  def self.cmd(cmd=nil)
    return @cmd if cmd.nil?
    @cmd = cmd
  end

  def self.desc(desc=nil)
    return @desc if desc.nil?
    @desc = desc
  end

  def self.parent_class(parent=nil)
    if parent.nil? && @parent.nil?
      YourMainCommand
    elsif parent.nil?
      @parent
    else
      @parent = parent
    end
  end
```

And plugins would look like:

```
class FooPlugin < PluginCommand
  cmd 'hello'
  desc 'say hello'
  parent_class SpeakCommand

  parameter "NAME", "Your name"

  def execute
    puts "Hello, #{name}"
  end
end

**Comment by mikeburr** [comment]
@kke Finding and loading the subcommands was relatively straightforward, especially since I don't have a need to handle nested subcommands.

From what I can tell, `subcommand` serves 3 purposes:

1. It lets Clamp include the subcommand in the help text.
2. It lets the Clamp parsing logic recognize the subcommand.
3. It tells Clamp to which class it should delegate parsing and execution of the subcommand.

In my particular scenario, (1) turns out to be undesirable (esp as the number of subcommands gets large) and `subcommand_missing` handles (2) and (3) nicely. 

I was actually quite surprised when I saw `subcommand_missing` in 1.1 as the implementation was dangerously close to the hack I was already using. I opened this bug mainly to inquire about the `has_subcommands?` behavior and the case of no statically-defined subcommands. It looks like the least painful path will be for me to define one or two subcommands statically (likely the one(s) that let the user list/search the subcommands) so my `subcommand_missing` will be invoked to find the others.
